### ISSUE TITLE
v1.3 docker-compile.sh error 

### ISSUE THREAD

**Issue by jykesaviaGNH** [apluslms/a-plus-rst-tools#117]
Strange error with `v1.3_stable` where the order of rst-file processing in Sphix can generate an error.
`Exception occurred:
  File "/usr/local/lib/python3.5/dist-packages/sphinx/environment/__init__.py", line 927, in get_doctree
    with open(doctree_filename, 'rb') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/compile/_build/doctrees/index.doctree' `

Steps to reproduce using the attached rst-course source in linux. It's important to start from clean build (`rm -rf _build`).

1. unzip `course.zip`
2. `cd t`
3. `git init`
4. `git submodule add https://github.com/apluslms/a-plus-rst-tools.git`
5. `(cd a-plus-rst-tools; git checkout v1.2_stable`)
6. `rm -rf _build; ./docker-compile.sh`  will work (warning about `eka_fi.rst` not included in any toctree)
7. `(cd a-plus-rst-tools; git checkout v1.3_stable)`  
8. `rm -rf _build; ./docker-compile.sh`   will **not** work
9. `mv esseet t_esseet`
10. `rm -rf _build; ./docker-compile.sh` will now work (`index.rst` is processed before `eka_fi.rst`

[course.zip](https://github.com/apluslms/a-plus-rst-tools/files/5791272/course.zip)

**Comment by Mankro** [comment]
Sounds like a severe problem. We will try to investigate this at some point. Thanks for reporting!

**Comment by Mankro** [comment]
Is this caused by the alphabetical order of filenames? If some chapter or module index file is compiled before the root index, then does it crash? Sphinx seems to compile files in the alphabetical order.

I wonder if this only affects multilingual courses.

**Comment by atilante** [comment]
I encountered the same at Aalto course CS-A1141, which is also multilingual. I am currently investigating the issue and hope to fix the bug.

**Comment by atilante** [comment]
### Quick fix

This problem seems to appear on **multilingual** courses.

Add the following lines to the **conf.py** file of the course:
```
enable_doc_link_multilang_suffix_correction = False
enable_ref_link_multilang_suffix_correction = False
```

This was mentioned in the commit message of 2a7f704445fdd9ecf2ad06f262dff4e5583ad42d .
This fixes both @jykesaviaGNH 's and my problem.

### Investigation

@jykesaviaGNH 's course in the attachment has the following file:
```
esseet/eka_fi.rst
```

Now, when the user runs the usual course compilation with `./docker-compile.sh`, the following happens.

1. Sphinx begins to read the RST files in the course directory in alphabetical order.
2. Sphinx encounters the file `esseet/eka_fi.rst` and triggers the event `source-read`.
3. In [aplus_setup.py of A+ RST tools](https://github.com/apluslms/a-plus-rst-tools/blob/master/aplus_setup.py#L54-L55), the event `source-read` is mapped to function `add_suffix_to_links`.
4. In A+ RST tools' file `toc_config.py`, the function [`add_suffix_to_links` activates.](https://github.com/apluslms/a-plus-rst-tools/blob/master/toc_config.py#L57)
5. Function add_suffix_to_links detects the language suffix `_fi` in the filename `esseet/eka_fi.rst`. [source](https://github.com/apluslms/a-plus-rst-tools/blob/master/toc_config.py#L83-L86)
6. Function add_suffix_to_links calls function [_is_multilingual_course()](https://github.com/apluslms/a-plus-rst-tools/blob/master/toc_config.py#L89) to determine whether the course is multilingual.
7. Function _is_multilingual_course [tries to read the Sphinx documentation tree](https://github.com/apluslms/a-plus-rst-tools/blob/master/toc_config.py#L52) to determine multilinguality.
8. Because Sphinx reads the RST files in alphabetical order, it has not yet processed the file `index.rst`. Therefore `get_doctree` raises a FileNotFoundError on file `/compile/_build/doctrees/index.doctree` (or something similar).

The same thing happens in the Aalto CS-A1141 course. There are two files:
```
esseet/aiheen_valinta.rst
esseet/aiheen_valinta_en.rst
```
The first one passes the compilation, because A+ RST tools does not detect a language tag in the filename. The compilation stops to an exception with the second file.

### Discussion
It seems that both @jykesaviaGNH and Ari Korhonen, the CS-A1141 course author, have had some RST files on their courses which have some language suffixes. This had been working before the new feature introduced in A+ RST tools v1.3 in commit 2a7f704445fdd9ecf2ad06f262dff4e5583ad42d  . Thus new A+ RST tools is smarter and tries to do nice things by default, but here it, unfortunately, produced quite a mystic compile error.

As a sidenode, I could fix the compilation problem on the CS-A1141 course **in another way** when I renamed the two RST files according to the multilingual directory structure that is used on the rest of the course, like this:

```
esseet/fi/aiheen_valinta.rst
esseet/en/aiheen_valinta.rst
```

### Proposals
1. Have the A+ RST tools settings `enable_doc_link_multilang_suffix_correction` and `enable_ref_link_multilang_suffix_correction` set to `False` by default.
2. Implement [`_is_multilingual_course` in **toc_config.py**](https://github.com/apluslms/a-plus-rst-tools/blob/master/toc_config.py#L51-L54) so that A+ RST tools can sniff the existence of a multilingual course on the fly, before Sphinx has processed the main `index.rst`.

**Comment by Mankro** [comment]
@atilante Thanks for the comprehensive report! I will add this to my backlog so that I would fix this better.

**Comment by jsorva** [comment]
FWIW, I seem to have written the following in my course’s conf.py at some point of time:
```
# See https://github.com/apluslms/a-plus-rst-tools/commit/2a7f704445fdd9ecf2ad06f262dff4e5583ad42d
# This is disabled, because the a-plus-rst-tools doc/ref-link mangling feature (in the above commit)
# 1) assumes that different language versions want to use the same link anchor (which results in none-too-pretty multilingual URLs like https://plus.cs.aalto.fi/o1/2020/w01/ch03/?hl=fi#shapeintro)
# 2) doesn’t (at least currently) seem to work with the ".. anchor::" shorthand defined in o1-rst-tools?
enable_doc_link_multilang_suffix_correction = False
enable_ref_link_multilang_suffix_correction = False

```

**Comment by Mankro** [comment]
This has been fixed in the v1.4_stable and v1.3_stable branches.
